api/types.go

```diff
@@ -515,19 +518,65 @@ func StringsToPeers(strs []string) []peer.ID {
 	return peers
 }
 
+// CidsToStrings encodes cid.Cids to strings.
+func CidsToStrings(cids []*cid.Cid) []string {
+	strs := make([]string, len(cids))
+	for i, c := range cids {
+		strs[i] = c.String()
+	}
+	return strs
+}
+
+// StringsToCids decodes cid.Cids from strings.
+func StringsToCids(strs []string) []*cid.Cid {
+	cids := make([]*cid.Cid, len(strs))
+	var err error
+	for i, str := range strs {
+		cids[i], err = cid.Decode(str)
+		if err != nil {
+			logger.Error(str, err)
+		}
+	}
+	return cids
+}
+
+// PinType values
+const (
+	DataType = iota
+	MetaType
+	CdagType
+	ShardType
+)
+
+// PinType specifies which of four possible interpretations a pin represents.
+// DataType pins are the simplest and represent a pin in the pinset used to
+// directly track user data.  ShardType pins are metadata pins that track
+// many nodes in a user's data DAG.  ShardType pins have a parent pin, and in
+// general can have many parents.  ClusterDAG, or Cdag for short, pins are also
+// metadata pins that do not directly track user data DAGs but rather other
+// metadata pins.  CdagType pins have at least one parent.  Finally MetaType
+// pins always track the cid of the root of a user-tracked data DAG.  However
+// MetaType pins are not stored directly in the ipfs pinset.  Instead the
+// underlying DAG is tracked via the metadata pins underneath the root of a
+// CdagType pin
+type PinType int
+
 // Pin is an argument that carries a Cid. It may carry more things in the
 // future.
 type Pin struct {
 	Cid                  *cid.Cid
 	Name                 string
+	Type                 PinType
 	Allocations          []peer.ID
 	ReplicationFactorMin int
 	ReplicationFactorMax int
 	Recursive            bool
+	Parents              []*cid.Cid
+	Child                *cid.Cid
```

Comment:
Maybe `Cdag` ? For consistency? It makes me wonder if `cdag` is a good name, or we should write `clusterdag` in full.